processor/src/main/java/com/linecorp/decaton/processor/processors/BatchingProcessor.java

```diff
@@ -0,0 +1,138 @@
+        public final T task;
+
+        private BatchingTask(Completion completion, ProcessingContext<T> context, T task) {
+            this.completion = completion;
+            this.context = context;
+            this.task = task;
+        }
+    }
+
+    /**
+     * Instantiate {@link BatchingProcessor}.
+     * If you only need one limit, please set large enough value to another.
+     * @param lingerMillis time limit for this processor. On every lingerMillis milliseconds,
+     * tasks in past lingerMillis milliseconds are pushed to {@link BatchingTask#processBatchingTasks(List)}.
+     * @param capacity size limit for this processor. Every time tasks’size reaches capacity,
+     * tasks in past before reaching capacity are pushed to {@link BatchingTask#processBatchingTasks(List)}.
+     */
+    public BatchingProcessor(long lingerMillis, int capacity) {
+        this.lingerMillis = lingerMillis;
+        this.capacity = capacity;
+
+        ScheduledThreadPoolExecutor scheduledExecutor = new ScheduledThreadPoolExecutor(1, r -> {
+                Thread th = new Thread(r);
+                th.setName("Decaton" + BatchingProcessor.class.getSimpleName()
+                           + '/' + System.identityHashCode(this));
+                return th;
+            });
+
+        // For faster shutdown cancel all pending flush on executor shutdown.
+        // In fact for this purpose we have two options here:
+        // A. Cancel all pending flush
+        // B. Run all pending flush immediately, regardless remaining time until flush
+        // A has a downside that it forces consumer to re-process some amount of tasks.
+        // B has a downside that it may causes flush bursting and then on spike for storage/API access in
+        // downstream processor.
+        // Given that this feature is expected to be used mainly for reducing workload for storage/APIs,
+        // we think taking A is much desirable behavior in most cases.
+        scheduledExecutor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
+        executor = scheduledExecutor;
+
+        scheduleFlush();
+    }
+
+    private void flush() {
+        if (windowedTasks.isEmpty()) {
+            return;
+        }
+        synchronized (windowedTasks) {
+            processBatchingTasks(new ArrayList<>(windowedTasks));
+            windowedTasks.clear();
+        }
+    }
+
+    private Runnable periodicallyFlushTask() {
+        return () -> {
+            flush();
+            scheduleFlush();
+        };
+    }
+
+    private void scheduleFlush() {
+        executor.schedule(periodicallyFlushTask(), lingerMillis, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void process(ProcessingContext<T> context, T task) throws InterruptedException {
+        BatchingTask<T> newTask = new BatchingTask<>(context.deferCompletion(), context, task);
+        windowedTasks.add(newTask);
+        if (windowedTasks.size() >= this.capacity) {
+            flush();
```

Comment:
Hm, then the problem is at the time `flush()` called, windowedTasks may contain more tasks than batch size, it's also not desirable.

Ok, then what do you think about following strategy?
- Instead of using single final `windowedTasks`, have non-final `List<BatchingTask>` for "current" batch.
- When "current" batch exceeds batch size, submit it to flusher and substitute current batch with new List.

Decaton will pause consumption if there's too many pending tasks, so even with this strategy, heap usage will not grow indefinitely.

Reply:
I'm sorry, I probably don't understand what you're saying...
I tried to correct it with my own understanding, but please point out if it is wrong 🙇 
cb78a8b